### PR TITLE
refactor(formanswer): supports generated targets without relation table

### PR DESCRIPTION
**Pull request by btry**
optimize a method

### CODE REVIEW

**Review by AdrienClairembault — DISMISSED** [review]

File: inc/abstractitiltarget.class.php [hunk 1]
```diff
@@ -96,9 +96,9 @@ abstract protected function getItem_Supplier();
     * Get an instance object for the relation between the target itemtype
     * and an object of any itemtype
     *
-    * @return CommonDBTM
+    * @return CommonDBRelation
```
Comment (btry) [inline inc/abstractitiltarget.class.php:99]: more precise type hinting

File: inc/abstractitiltarget.class.php [hunk 2]
```diff
@@ -96,9 +96,9 @@ abstract protected function getItem_Supplier();
...
     */
-   abstract public function getItem_Item();
+   abstract public static function getItem_Item(): CommonDBRelation;
```
Comment (btry) [inline inc/abstractitiltarget.class.php:101]: switching to static method to avoif creating an object for this call which does not uses $this

File: inc/abstractitiltarget.class.php [hunk 3]
```diff
@@ -2240,4 +2240,43 @@ protected function prepareUploadedFiles(array $data, $formanswer): array {
 
       return $data;
    }
+
+   public function isObjectTarget(): bool {
+      return true;
+   }
```
Comment (btry) [inline inc/abstractitiltarget.class.php:2246]: provide info about the task of this target : it generates an object

File: inc/abstractitiltarget.class.php [hunk 4]
```diff
@@ -2240,4 +2240,49 @@ protected function prepareUploadedFiles(array $data, $formanswer): array {
...
+
+   /**
+    * Find generated targets for this target and the given form answer
+    *
+    * @param PluginFormcreatorFormAnswer $formAnswer
+    * @return array
+    */
+   public static function findForFormAnswer(PluginFormcreatorFormAnswer $formAnswer): array {
```
Comment (btry) [inline inc/abstractitiltarget.class.php:2254]: this method comes from the move of the SQL query from PluginFormcreatorFormAnswer::getGeneratedTargets()

File: inc/formanswer.class.php [hunk 5]
```diff
@@ -862,13 +863,16 @@ public function generateTarget(): bool {
...
-               $success = false;
-               break;
+            if ($targetObject->isObjectTarget()) {
+               if ($generatedTarget === null) {
+                  // If the target generates an object but none generated
+                  $success = false;
+               } else {
+                  $this->targetList[] = $generatedTarget;
+                  // Map [itemtype of the target] [item ID of the target] = ID of the generated target
+                  $generatedTargets->addTarget($targetObject, $generatedTarget);
```
Comment (btry) [inline inc/formanswer.class.php:873]: populate targetList only for generataed objects, and feed the composite tickets linker only in this case.

File: inc/formanswer.class.php [hunk 6]
```diff
@@ -44,6 +44,7 @@ class PluginFormcreatorFormAnswer extends CommonDBTM
 
    /**
     * Generated targets after creation of a form answer
+    * Populated only after creation of a FormAnswer in DB with add() method
```
Comment (btry) [inline inc/formanswer.class.php:47]: this property is used only to completesmart HTTP redirection after ceating a formanswer from a form. See front/form.form.php

File: inc/formanswer.class.php [hunk 7]
```diff
@@ -1828,15 +1832,11 @@ public function  getCurrentApprovers(): ?array {
     * get all generated targets by the form answer
     * populates the generated targets associated to the instance
     *
-    * Assumes that the generated target has a relation with the form answer
```
Comment (btry) [inline inc/formanswer.class.php:1831]: No longer true

File: inc/formanswer.class.php [hunk 8]
```diff
@@ -1848,44 +1848,8 @@ public function getGeneratedTargets($itemtypes = []): array {
          $itemtypes = array_intersect(PluginFormcreatorForm::getTargetTypes(), $itemtypes);
       }
       /** @var PluginFormcreatorTargetInterface $targetType */
-      $this->targetList = [];
-      foreach (PluginFormcreatorForm::getTargetTypes() as $targetType) {
```
Comment (btry) [inline inc/formanswer.class.php:1852]: moved in AbstractItilTarget

File: inc/targetinterface.class.php [hunk 9]
```diff
@@ -37,5 +37,27 @@ interface PluginFormcreatorTargetInterface
...
+    */
+   public function isObjectTarget(): bool;
+
+   /**
+    * Find targets linked to a form answer
+    *
+    * @param PluginFormcreatorFormAnswer $formAnswer
+    * @return array
+    */
+   public static function findForFormAnswer(PluginFormcreatorFormAnswer $formAnswer): array;
```
Comment (btry) [inline inc/targetinterface.class.php:62]: 3 new mandatory methods

File: tests/3-unit/PluginFormcreatorFormAnswer.php [hunk 10]
```diff
@@ -699,23 +699,16 @@ public function testGetGeneratedTargets() {
       $generatedTargets = $instance->targetList;
 
       $output = $instance->getGeneratedTargets();
-      // Check the targetsList is unchanged
       $this->array($instance->targetList)->hasSize(count($generatedTargets));
       $this->array($output)->hasSize(3);
 
       $output = $instance->getGeneratedTargets([PluginFormcreatorTargetTicket::getType()]);
-      // Check the targetsList is unchanged
-      $this->array($instance->targetList)->hasSize(3);
```
Comment (btry) [inline tests/3-unit/PluginFormcreatorFormAnswer.php:708]: getGeneratedObjects does no longer populate targetList : useless.

File: inc/targetinterface.class.php [hunk 11]
```diff
@@ -37,5 +37,27 @@ interface PluginFormcreatorTargetInterface
...
+    * @return boolean true if the target does an action
+    */
+   public function isActionTarget(): bool;
+
+   /**
+    * Tell if this target generates an object and does not does an action
+    *
+    * @return boolean true if the target generates an object
+    */
+   public function isObjectTarget(): bool;
```
Comment (AdrienClairembault) [inline inc/targetinterface.class.php:54]: The comments seems to indicate that `isActionTarget` and `isObjectTarget` are exclusive.
Wouldn't a single method `getTargetType` returning something like `AbstractTarget::TYPE_ACTION` or `AbstractTarget::TYPE_OBJECT` be better ?

`isActionTarget` and `isObjectTarget` could stays as helper methods in the base abstract class but only `getTargetType` would be required by the interface.

File: inc/targetinterface.class.php [hunk 12]
```diff
@@ -37,5 +37,27 @@ interface PluginFormcreatorTargetInterface
...
 
-   public function getTargetItemtypeName();
+   public static function getTargetItemtypeName();
+
+   /**
+    * Tell if this target does an action and does not generate an object
+    *
+    * @return boolean true if the target does an action
+    */
+   public function isActionTarget(): bool;
```
Comment (AdrienClairembault) [inline inc/targetinterface.class.php:47]: You set multiple similar functions as static in this PR, maybe this one should be static too ?

File: inc/targetinterface.class.php [hunk 13]
```diff
@@ -37,5 +37,27 @@ interface PluginFormcreatorTargetInterface
...
+
+   /**
+    * Tell if this target generates an object and does not does an action
```
Comment (AdrienClairembault) [inline inc/targetinterface.class.php:50]:
```suggestion
    * Tell if this target generates an object and does not do an action
```

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
Looks good overall expected for a few minor changes (see comments above).

If I get it correctly, targets like `EmailTarget` from advanced forms does not count as generating "objects".
Even if `EmailTarget` create a `SentEmail` object, since this object is only used to confirm that the action was run properly, it is not expected to be returned by `findForFormAnswer`.

File: inc/targetinterface.class.php (hunk 11 above)
Comment (btry) [reply inc/targetinterface.class.php:54]: > The comments seems to indicate that isActionTarget and isObjectTarget are exclusive

Yes I consider they are exclusive. Actually creating a ticket is already an action by itself, but it creates an object on the DB. Your SentEmail is also an object, then EmailTarget generates an object. 

> The comments seems to indicate that isActionTarget and isObjectTarget are exclusive.
Wouldn't a single method getTargetType returning something like AbstractTarget::TYPE_ACTION or AbstractTarget::TYPE_OBJECT be better ?
>
> **isActionTarget** and **isObjectTarget** could stays as helper methods in the base abstract class but only getTargetType would be required by the interface.

If I switch to a single method returning a constant, then I wwould drop **isActionTarget** and **isObjectTarget**  from this PR.

File: inc/targetinterface.class.php (hunk 12 above)
Comment (btry) [reply inc/targetinterface.class.php:47]: I agree, but not in this PR. And I focus on releasing an Alpha 4. I wanted to release it last week already.

**Review by AdrienClairembault — APPROVED** [review]

**Pull request merged**